### PR TITLE
[coding] Added a version with exceptions to MemReader.

### PR DESCRIPTION
**Pull request by mpimenov**

### CODE REVIEW

File: coding/reader.hpp [hunk 1]
```diff
@@ -50,27 +49,55 @@ class MemReader : public Reader
...
-    ASSERT ( AssertPosAndSize(pos, size), () );
-    return make_unique<MemReader>(m_pData + pos, static_cast<size_t>(size));
+    AssertPosAndSize(pos, size);
+    return make_unique<MemReaderTemplate>(m_pData + pos, static_cast<size_t>(size));
   }
 
 private:
+  // We assume that this check is not compiled into
+  // the release version of MemReaderTemplate<false>.
+  bool GoodPosAndSize(uint64_t pos, uint64_t size) const
```
Comment (ygorshenin) [inline coding/reader.hpp]: Even if it's compiled, it's not a problem.

File: coding/reader.hpp [hunk 2]
```diff
@@ -50,27 +49,55 @@ class MemReader : public Reader
...
+    bool const ret2 = (size <= static_cast<size_t>(-1));
+    return ret1 && ret2;
+  }
+
+  void AssertPosAndSize(uint64_t pos, uint64_t size) const
+  {
+    if (WithExceptions)
+    {
+      if (!GoodPosAndSize(pos, size))
+      {
```
Comment (ygorshenin) [inline coding/reader.hpp]: No need in curly braces.

**Comment by mpimenov** [comment]
PTAL

File: traffic/traffic_info.cpp [hunk 3]
```diff
@@ -264,7 +264,7 @@ void TrafficInfo::SerializeTrafficKeys(vector<RoadSegmentId> const & keys, vecto
 void TrafficInfo::DeserializeTrafficKeys(vector<uint8_t> const & data,
                                          vector<TrafficInfo::RoadSegmentId> & result)
 {
-  MemReader memReader(data.data(), data.size());
+  MemReaderWithExceptions memReader(data.data(), data.size());
```
Comment (ygorshenin) [inline traffic/traffic_info.cpp]: Where do you plan to catch possible exceptions?

Comment (mpimenov) [inline traffic/traffic_info.cpp]: done

**Comment by mpimenov** [comment]
PTAL

File: traffic/traffic_info.cpp [hunk 4]
```diff
@@ -107,7 +107,21 @@ TrafficInfo::TrafficInfo(MwmSet::MwmId const & mwmId, int64_t currentDataVersion
...
       reader.Read(0, buf.data(), buf.size());
       LOG(LINFO, ("Reading keys for", mwmId, "from section"));
-      DeserializeTrafficKeys(buf, m_keys);
+      try
+      {
+        DeserializeTrafficKeys(buf, m_keys);
+      }
+      catch (Reader::Exception const & e)
+      {
+        auto const info = mwmId.GetInfo();
```
Comment (ygorshenin) [inline traffic/traffic_info.cpp]: What if mwmId is invalid?

Comment (mpimenov) [inline traffic/traffic_info.cpp]: Added safety checks.

**Comment by ygorshenin** [comment]
LGTM

**Pull request merged**